skfda/misc/metrics/_mahalanobis.py

```diff
@@ -0,0 +1,143 @@
+        alpha: Hyperparameter that controls the smoothness of the
+            aproximation.
+        eigenvectors: Eigenvectors of the covariance operator.
+        eigenvalues: Eigenvalues of the covariance operator.
+
+    Examples:
+        >>> from skfda.misc.metrics import MahalanobisDistance
+        >>> import numpy as np
+        >>> from skfda.representation.grid import FDataGrid
+        >>> data_matrix = np.array([[1.0, 0.0], [0.0, 2.0]])
+        >>> grid_points = [0, 1]
+        >>> fd = FDataGrid(data_matrix, grid_points)
+        >>> mahalanobis = MahalanobisDistance(2)
+        >>> mahalanobis.fit(fd)
+        MahalanobisDistance(n_components=2)
+        >>> mahalanobis(fd[0], fd[1])
+        array([ 1.99680384])
+
+    References:
+        .. footbibliography::
+    """
+
+    def __init__(
+        self,
+        n_components: int = 10,
+        centering: bool = True,
+        regularization: Optional[TikhonovRegularization[FData]] = None,
+        weights: Optional[Union[ArrayLike, WeightsCallable]] = None,
+        components_basis: Optional[Basis] = None,
+        alpha: float = 0.001,
+        eigenvalues: Optional[NDArrayFloat] = None,
+        eigenvectors: Optional[FData] = None,
+    ) -> None:
+        self.n_components = n_components
+        self.centering = centering
+        self.regularization = regularization
+        self.weights = weights
+        self.components_basis = components_basis
+        self.alpha = alpha
+        self.eigenvalues = eigenvalues
+        self.eigenvectors = eigenvectors
+
+    def fit(
+        self,
+        X: FData,
+        y: None = None,
+    ) -> MahalanobisDistance:
+        """Fit the functional Mahalanobis distance to X.
+
+        We extract the eigenvectors and corresponding eigenvalues
+        of the covariance operator by using FPCA.
+
+        Args:
+            X: Stochastic process.
+            y: Ignored.
+
+        Returns:
+            self
+        """
+        from ...preprocessing.dim_reduction import FPCA
+
+        if self.eigenvalues is None or self.eigenvectors is None:
+            fpca = FPCA(
+                self.n_components,
+                self.centering,
+                self.regularization,
+                self.weights,
+                self.components_basis,
+            )
+            fpca.fit(X)
+            self.eigenvalues_ = fpca.explained_variance_
+            self.eigenvectors_ = fpca.components_
+
+        return self
+
+    def __call__(
+        self,
+        e1: FData,
+        e2: FData,
+    ) -> NDArrayFloat:
+        """Compute the squared functional Mahalanobis distances of given observations.
+
+        Args:
+            e1: First object.
+            e2: Second object.
+
+        Returns:
+            Squared functional Mahalanobis distance between two observations.
+        """
+        if self.eigenvalues is not None and self.eigenvectors is not None:
+            self.eigenvalues_ = self.eigenvalues
```

Comment:
This should be done in `fit`. If `fit` is not called, then you can use a local variable to get either `self.eigenvalues_` or `self.eigenvalues`, but you shouldn't change the object in `__call__`.